### ISSUE TITLE
Interface with optional properties incorrectly allows setting other string properties

### ISSUE THREAD

**Issue by rdhelms** [microsoft/TypeScript#35628]
**TypeScript Version:**  3.7.2

**Search Terms:**
Interface with optional properties incorrectly allows setting other string properties

**Code**

```ts
interface IAnyKeys {
    [key: string]: number;
}

interface ISpecificKeys {
    someProp?: number;
}

type Test = IAnyKeys extends ISpecificKeys ? true : false;   // true, but expected false

const thing: ISpecificKeys = {
    ['randomProp' as string]: 123    // This is allowed?
}
```

**Expected behavior:**
Expected `IAnyKeys` not to extend `ISpecificKeys`, and expected a type error when trying to set a string property on the object typed as `ISpecificKeys`.

**Actual behavior:**
`IAnyKeys` extends `ISpecificKeys` and setting a string-typed property on the `ISpecificKeys`-typed object is allowed.

**Playground Link:** 
https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgJIEEQE8DSEsDOyA3gFDIXIDaA1vgFzIFhSgDmAuoyAK4C2AI2gBuUgF9SpUJFiIUqAMoAHCAmAxgCPIRLlKBAPZ8IABSgGlAfm78hUURNJgsK5ABUIzZAF40mXPhEEAAekCAAJkSKKmoaWoHIlsgsPCiM8AA2BBDCyAD0ecghMZDhyJnZkggGIF5gABbsjNGq6praRL5klNQA5FBwEUZmFr3IcETMrCCcjACMAEwAzD0F7o1EwERwGRkGAO4Q4ZbipEA

**Related Issues:** 
Possibly https://github.com/microsoft/TypeScript/pull/26797 ...?

**Comment by RyanCavanaugh** [comment]
`Test` is simply a case of incorrect expectation. `true` is the right result because reads from `ISpecificKeys` do result in `number`.

`thing` is the same thing described in #35538

**Comment by fatcerberus** [comment]
> reads from ISpecificKeys do result in number.

Not necessarily though - even assuming TS is only looking at known properties, there is an optional property in `ISpecificKeys` and so you might read `undefined` - which isn’t assignable to `number`?

**Comment by RyanCavanaugh** [comment]
Reads of present properties 😉

**Comment by fatcerberus** [comment]
Sorry, I should have been more specific.  I meant to point to this discrepancy:
```ts
const x: { foo?: number } = {};
const y: { [x: string]: number } = {};
x.foo = undefined;  // ok, no error
y.foo = undefined;  // error!
```

The index signature isn't considered to have `undefined` in its domain.  The optional property does.

**Comment by rdhelms** [comment]
Interesting - so if this is true (from #35538)...

> When the name isn't statically determinable, we error on the lenient side and assume you're using some construction that will resolve to a declared property.

...then why does an attempt to set a string property fail if it's after the initial assignment?

```ts
interface IAnyKeys {
    [key: string]: number;
}

interface ISpecificKeys {
    someProp?: number;
}

type Test = IAnyKeys extends ISpecificKeys ? true : false;   // true, but expected false

const thing: ISpecificKeys = {
    ['randomProp' as string]: 123    // This is allowed, as above
}

thing['randomProp' as string] = 123;    // But this is an error?
```

I thought these two cases would have similar type checking, but I'd love to be enlightened as to why they're different.

**Comment by rdhelms** [comment]
Also, it's possibly worth pointing out that the types of the index signature and `someProp` both being `number` doesn't seem to matter. `someProp` can have any type whatsoever (so long as it's optional) and `Test` will still be `true`:

```ts
interface IAnyKeys {
    [key: string]: number;
}

interface ISpecificKeys {
    someProp?: string;
}

type Test = IAnyKeys extends ISpecificKeys ? true : false;     // true, but expected false
```

**Comment by RyanCavanaugh** [comment]
```ts
const thing: ISpecificKeys = {
    ['randomProp' as string]: 123    // This is allowed, as above
}

thing['randomProp' as string] = 123;    // But this is an error?
```

In the first case, we're checking whether or not the initializer is a valid `ISpecificKeys`. We can tell there's *some* property, but we don't know its name. `ISpecificKeys` doesn't require *anything*`, `{}` is a valid `ISpecificKeys`, so `{} along with some other property` *must* also be valid because adding a (non-conflicting) property is an allowed subtyping relation.

In the second case, it's an arbitrary property assignment, and you didn't say that `ISpecificKeys` was the kind of type that allowed for arbitrary property assignment (which you would do by writing an index signature).

This is similar to how this code is OK
```ts
const a = { x: 0, y: 0 };
// Valid
const b: { x: number } = a;
``
but not this
```ts
const a = { x: 0, y: 0 };
const b: { x: number } = a;
// Not valid
b.y = 0;
```

**Comment by rdhelms** [comment]
Thanks @RyanCavanaugh - I appreciate the clarification. Honestly, I think this was the source of confusion for me:
> adding a (non-conflicting) property is an allowed subtyping relation.

I think that especially is confusing when something like this is *not* allowed:
```ts
function test(thing: { x: number }) { }
test({ x: 0, y: 0 })    // Error
```

Also, is there a chance you could provide some clarification on why https://github.com/microsoft/TypeScript/issues/35628#issuecomment-565116369 is still the case? Based on the explanation that "reads from `ISpecificKeys` do result in number", I was expecting changing `someProp` from `number` to `string` would result in `Test` being `false`

**Comment by fatcerberus** [comment]
The reason `test()` above is an error is because of excess property checks in object literals.  If you’re coming up with an object on-the-fly, it’s most likely an error if you provide properties the type doesn’t call for, so the compiler calls you out for it.  But in general, TypeScript is structurally typed—so “that thing, but with extra properties” is actually the *only* subtyping relationship that matters.

This is a good explanation of how that all fits together:
https://basarat.gitbooks.io/typescript/docs/types/freshness.html#freshness
